File: ts/test/token-class.test.js
/* Copyright (c) 2026 Richard Rodger and other contributors, MIT License */
'use strict'

// `tokenClasses` compiles a production whose alternatives are all single
// literals or engine tokens to one engine token set. What a grammar
// accepts must not depend on the option: each case here is one where it
// once did (tabnas/bnf#74 review). Go and Rust pin the same cases.

const { describe, it } = require('node:test')
const assert = require('node:assert')

const { emitGrammarSpec } = require('../dist/bnf')
const { Tabnas } = require('@tabnas/parser')

const lit = (s) => ({ kind: 'term', literal: s, caseSensitive: true })
const rx = (pattern) => ({ kind: 'regex', pattern, flags: '' })
const tok = (name) => ({ kind: 'token', name })
const ref = (name) => ({ kind: 'ref', name })
const prod = (name, ...alts) => ({ name, alts })

const emit = (productions, tokenClasses) =>
  emitGrammarSpec({ productions }, { tag: 'tc', start: 'doc', tokenClasses })
const parses = (spec, src, opts) => {
  try {
    return new Tabnas(opts || {}).grammar(spec).parse(src).rule === 'doc'
  } catch (e) {
    return false
  }
}
const seqs = (spec, name) => spec.rule[name].open.map((o) => o.s)
const sets = (spec) => Object.keys(spec.options.tokenSet || {})
const relex = { lex: { relex: true } }

describe('token-class', () => {

  it('a class set head takes the keyword-shadow order its members had', () => {
    // doc = x ; x = C / [a-z] D ; C = "a" / "b" ; D = [0-9]
    // With the option off C's literals are heads of x, each guarded ahead
    // of the character class that shadows it. The set that stands for
    // them is ordered the same way, or \`a1\` commits to C at \`a\` and
    // fails at \`1\`.
    const g = [
      prod('doc', [ref('x')]),
      prod('x', [ref('C')], [rx('[a-z]'), ref('D')]),
      prod('C', [lit('a')], [lit('b')]),
      prod('D', [rx('[0-9]')]),
    ]
    const off = emit(g, false)
    const on = emit(g, true)
    assert.deepEqual(seqs(on, 'x'), ['#C #ZZ', '#RX__A_Z', '#C'])
    for (const src of ['a1', 'a', 'b', 'c1', 'b1']) {
      assert.ok(parses(off, src, relex), 'off: ' + src)
      assert.ok(parses(on, src, relex), 'on: ' + src)
    }
  })

  it('a class with a set among its members stays a production', () => {
    // C = #C / "a" names its own set; C = #D / "a" beside D = #C / "b"
    // names another class's. A set of sets is one the engine cannot
    // resolve, and expanding it never ended.
    const x = prod('x', [ref('C'), ref('t'), ref('u')], [lit('z'), ref('u'), ref('t')])
    const rest = [prod('t', [lit('.'), lit('.')]), prod('u', [lit(','), lit(',')])]
    for (const classes of [
      [prod('C', [tok('#C')], [lit('a')])],
      [prod('C', [tok('#D')], [lit('a')]), prod('D', [tok('#C')], [lit('b')])],
    ]) {
      const g = [prod('doc', [ref('x')]), x, ...rest, ...classes]
      const on = emit(g, true)
      assert.deepEqual(sets(on), [])
      assert.deepEqual(seqs(on, 'x'), seqs(emit(g, false), 'x'))
      assert.ok(parses(on, 'a..,,'))
      assert.ok(parses(on, 'z,,..'))
    }
  })

  it('an empty production name is not a class', () => {
    // doc = x ; x = <""> "!" ; <""> = "a" / "b". The set would be named
    // \`#\`, which names nothing, and so would the token standing for the
    // reference.
    const g = [
      prod('doc', [ref('x')]),
      prod('x', [ref(''), lit('!')]),
      prod('', [lit('a')], [lit('b')]),
    ]
    const on = emit(g, true)
    assert.deepEqual(sets(on), [])
    assert.deepEqual(seqs(on, 'x'), seqs(emit(g, false), 'x'))
    assert.ok(parses(on, 'a!'))
    assert.ok(parses(on, 'b!'))
  })


  it('a production with a member that consumes nothing is not a class', () => {
    // doc = x ; x = C "b" ; C = <member> / "a". An empty literal matches
    // nothing, and #ZZ and #AA can be satisfied without input, while the
    // set standing for a class is one token and never empty.
    for (const member of [lit(''), tok('#ZZ'), tok('#AA')]) {
      const g = [
        prod('doc', [ref('x')]),
        prod('x', [ref('C'), lit('b')]),
        prod('C', [member], [lit('a')]),
      ]
      const on = emit(g, true)
      assert.deepEqual(sets(on), [], JSON.stringify(member))
      assert.deepEqual(seqs(on, 'x'), seqs(emit(g, false), 'x'))
      assert.ok(parses(on, 'ab'))
    }
  })


  it('a production whose name holds whitespace is not a class', () => {
    // The set standing for a class is named after it, and an alternate's
    // `s` separates token names with whitespace, so `#C D` would read as
    // two tokens. Such a production stays plain, as with the option off.
    for (const name of ['C D', 'C\tD', 'C\u00a0D', 'C\u0085D', 'C\ufeffD']) {
      const g = [
        prod('doc', [ref('x')]),
        prod('x', [ref(name), lit('b')]),
        prod(name, [lit('a')], [lit('c')]),
      ]
      const on = emit(g, true)
      assert.deepEqual(sets(on), [], JSON.stringify(name))
      assert.deepEqual(seqs(on, 'x'), seqs(emit(g, false), 'x'))
      assert.ok(parses(on, 'ab'), JSON.stringify(name))
      assert.ok(parses(on, 'cb'), JSON.stringify(name))
    }
  })

})

File: rs/tests/token_class_test.rs
// Copyright (c) 2026 Richard Rodger and other contributors, MIT License

//! `token_classes` compiles a production whose alternatives are all
//! single literals or engine tokens to one engine token set. What a
//! grammar accepts must not depend on the option: each case here is one
//! where it once did (tabnas/bnf#74 review). Mirrors
//! `ts/test/token-class.test.js`.

mod common;

use common::{alt_seqs, prod, reference, rx, sens_term, tok, token_set_names};
use tabnas_bnf::{emit_grammar_spec, ConvertOptions, Grammar, GrammarSpec, Production};

fn emit(prods: Vec<Production>, token_classes: bool) -> GrammarSpec {
    let opts = ConvertOptions::tag("tc")
        .start("doc")
        .token_classes(token_classes);
    emit_grammar_spec(&Grammar::new(prods), &opts).expect("emit")
}

fn parses(spec: &GrammarSpec, src: &str, relex: bool) -> bool {
    let mut options = tabnas::Options::default();
    options.lex.relex = relex;
    let mut parser = tabnas::Tabnas::with_options(options);
    spec.install(&mut parser).expect("install");
    parser.parse(src).is_ok()
}

#[test]
fn a_class_set_head_takes_the_keyword_shadow_order_its_members_had() {
    // doc = x ; x = C / [a-z] D ; C = "a" / "b" ; D = [0-9]
    // With the option off C's literals are heads of x, each guarded ahead
    // of the character class that shadows it. The set that stands for
    // them is ordered the same way, or `a1` commits to C at `a` and fails
    // at `1`.
    let g = || {
        vec![
            prod("doc", vec![vec![reference("x")]]),
            prod(
                "x",
                vec![vec![reference("C")], vec![rx("[a-z]", ""), reference("D")]],
            ),
            prod("C", vec![vec![sens_term("a")], vec![sens_term("b")]]),
            prod("D", vec![vec![rx("[0-9]", "")]]),
        ]
    };
    let off = emit(g(), false);
    let on = emit(g(), true);
    assert_eq!(
        alt_seqs(&on, "x"),
        [
            Some("#C #ZZ".to_string()),
            Some("#RX__A_Z".to_string()),
            Some("#C".to_string())
        ]
    );
    for src in ["a1", "a", "b", "c1", "b1"] {
        assert!(parses(&off, src, true), "off: {src}");
        assert!(parses(&on, src, true), "on: {src}");
    }
}

#[test]
fn a_class_with_a_set_among_its_members_stays_a_production() {
    // C = #C / "a" names its own set; C = #D / "a" beside D = #C / "b"
    // names another class's. A set of sets is one the engine cannot
    // resolve, and expanding it never ended.
    let classes: [fn() -> Vec<Production>; 2] = [
        || vec![prod("C", vec![vec![tok("#C")], vec![sens_term("a")]])],
        || {
            vec![
                prod("C", vec![vec![tok("#D")], vec![sens_term("a")]]),
                prod("D", vec![vec![tok("#C")], vec![sens_term("b")]]),
            ]
        },
    ];
    for class in classes {
        let g = || {
            let mut out = vec![
                prod("doc", vec![vec![reference("x")]]),
                prod(
                    "x",
                    vec![
                        vec![reference("C"), reference("t"), reference("u")],
                        vec![sens_term("z"), reference("u"), reference("t")],
                    ],
                ),
                prod("t", vec![vec![sens_term("."), sens_term(".")]]),
                prod("u", vec![vec![sens_term(","), sens_term(",")]]),
            ];
            out.extend(class());
            out
        };
        let on = emit(g(), true);
        assert!(
            token_set_names(&on).is_empty(),
            "{:?}",
            token_set_names(&on)
        );
        assert_eq!(alt_seqs(&on, "x"), alt_seqs(&emit(g(), false), "x"));
        assert!(parses(&on, "a..,,", false));
        assert!(parses(&on, "z,,..", false));
    }
}

#[test]
fn an_empty_production_name_is_not_a_class() {
    // doc = x ; x = <""> "!" ; <""> = "a" / "b". The set would be named
    // `#`, which names nothing, and so would the token standing for the
    // reference.
    let g = || {
        vec![
            prod("doc", vec![vec![reference("x")]]),
            prod("x", vec![vec![reference(""), sens_term("!")]]),
            prod("", vec![vec![sens_term("a")], vec![sens_term("b")]]),
        ]
    };
    let on = emit(g(), true);
    assert!(
        token_set_names(&on).is_empty(),
        "{:?}",
        token_set_names(&on)
    );
    assert_eq!(alt_seqs(&on, "x"), alt_seqs(&emit(g(), false), "x"));
    assert!(parses(&on, "a!", false));
    assert!(parses(&on, "b!", false));
}

#[test]
fn a_production_with_a_member_that_consumes_nothing_is_not_a_class() {
    // doc = x ; x = C "b" ; C = <member> / "a". An empty literal matches
    // nothing, and #ZZ and #AA can be satisfied without input, while the
    // set standing for a class is one token and never empty.
    for member in [sens_term(""), tok("#ZZ"), tok("#AA")] {
        let g = || {
            vec![
                prod("doc", vec![vec![reference("x")]]),
                prod("x", vec![vec![reference("C"), sens_term("b")]]),
                prod("C", vec![vec![member.clone()], vec![sens_term("a")]]),
            ]
        };
        let on = emit(g(), true);
        assert!(
            token_set_names(&on).is_empty(),
            "{:?}",
            token_set_names(&on)
        );
        assert_eq!(alt_seqs(&on, "x"), alt_seqs(&emit(g(), false), "x"));
        assert!(parses(&on, "ab", false));
    }
}
